Add GitHub Actions workflow to run unit tests and check code coverage

per1234 opened the pull request:
On every pull request or push that changes relevant files (anything under `extras/test/` or `src/`):
- Run unit tests
- Print coverage report to the workflow log
- Upload coverage report to Codecov

This PR is dependent on https://github.com/arduino-libraries/ArduinoIoTCloud/pull/118. The workflow is currently expected to fail because the unit tests won't be added until https://github.com/arduino-libraries/ArduinoIoTCloud/pull/118 is merged.

Demonstration:
https://github.com/per1234/ArduinoIoTCloud/runs/646210541?check_suite_focus=true

https://codecov.io/github/per1234/ArduinoIoTCloud/commit/77fe90bb786f3e0843cb03b2839d1fd58c94995e

Fixes https://github.com/arduino-libraries/ArduinoIoTCloud/issues/119

aentinger commented:
Good morning :wave: Merging #122 in I created a small conflict on the README. Can you please rebase and resolve? I've also been thinking - as a matter of fact we can merge this now and I just rebase the #118 OTA branch. Another possibility would be to merge this directly into the OTA #118 branch. This has the added benefit of not showing a "broken-build" badge on the README as long as the OTA branch is not merged. What are your thoughts on this?

per1234 commented:
>Can you please rebase and resolve?

I have now done this.

>I've also been thinking - as a matter of fact we can merge this now and I just rebase the #118 OTA branch.

I wouldn't want the broken badge in the readme for a long period of time, but if it's only a short period, then no big deal.

>Another possibility would be to merge this directly into the OTA #118 branch. This has the added benefit of not showing a "broken-build" badge on the README as long as the OTA branch is not merged. What are your thoughts on this?

It's fine with me if you want it in that PR. That would cause it to be merged when it's needed. Since #118 is from a branch (rather than a fork as in this PR), I believe it will also allow the Unit Test workflow to run on the PR so you can verify the workflow is working correctly.

aentinger commented:
>It's fine with me if you want it in that PR. That would cause it to be merged when it's needed. Since #118 is from a branch (rather than a fork as in this PR), I believe it will also allow the Unit Test workflow to run on the PR so you can verify the workflow is working correctly.

That's great! Can you readjust this PR so it can be merged to https://github.com/arduino-libraries/ArduinoIoTCloud/tree/ota instead of master?

per1234 commented:
I have now changed the base branch to `ota`. It brought in https://github.com/arduino-libraries/ArduinoIoTCloud/pull/121/commits/aa1fb9520c2853e3ee370276d65b0d4ffef4032a, since I had rebased the PR on `master` after #122 was merged, but I figure #118 will need to rebase on that commit before merging anyway.

The "Check code coverage" step of the Unit Tests workflow is currently failing because CMakeLists.txt doesn't add the `--coverage` flags needed to output the coverage data:
```
set(CMAKE_C_FLAGS   ${CMAKE_C_FLAGS}   "--coverage")
set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} "--coverage")
```
Would you like me to add those lines to CMakeLists.txt?

aentinger commented:
>Would you like me to add those lines to CMakeLists.txt?

Yes please :wink: :+1:
